design/one-pager-merge-jets-proposal.md

```diff
@@ -0,0 +1,402 @@
+about its internals in [the deep dive blog post series][deep-dive-blogs].
+
+Before releasing the first Terrajet-based providers, we have had a discussion
+whether we'd like to have them as separate providers or have the
+Terrajet-generated CRDs added to the existing classic providers in the [provider
+strategy doc][strategy-doc]. The decision for initial releases of the first
+three Jet-based providers ([AWS][jet-aws], [Azure][jet-azure] and
+[GCP][jet-gcp]) was to have them as separate providers for the reasons stated
+[here](https://github.com/crossplane/crossplane/blob/master/design/design-doc-provider-strategy.md#decision-for-initial-releases).
+
+We called out that we'll revisit this decision after more usage and community
+feedback, which was lacking most probably because we were asking people (in
+Slack and in https://github.com/crossplane/crossplane/pull/2701) for their
+opinions about providers that they have not used yet. Today we have more data
+and more awareness in the community, and having two different implementations
+has been confusing for a lot of users, so a revisit of the decision regarding
+AWS, Azure and GCP is due.
+
+## Goals
+
+We have the following goals for _the big three_ providers:
+
+* The cloud vendors should maintain their Crossplane provider.
+* The users should have the ability to provision all of their infrastructure
+  through Crossplane, i.e. coverage.
+* The providers should be mature enough for users to depend on them in
+  production, i.e. maturity.
+
+Today, we have two Crossplane providers for each of the big three clouds.
+* AWS
+  * Classic with [124 CRDs][provider-aws-crds] with 37 beta CRDs.
+  * Jet with [763 CRDs][jet-aws-crds], no beta CRDs.
+* Azure
+  * Classic with [21 CRDs][provider-azure-crds] with 7 beta CRDs.
+  * Jet with [647 CRDs][jet-azure-crds], no beta CRDs.
+* Google Cloud
+  * Classic with [27 CRDs][provider-gcp-crds] with 11 beta CRDs.
+  * Jet with [438 CRDs][jet-gcp-crds], no beta CRDs.
+
+Both variants of each reach those goals at different levels but none of them are
+maintained by their cloud vendor. The aim of this document is to capture a
+decision about the future of these provider implementations with the goals
+stated above in mind.
+
+## Proposal
+
+There were three options in the provider strategy doc:
+* Option A: Completely separate providers.
+* Option B: One provider but Jet-based resources in their own API group.
+* Option C: One provider and no Jet vs classic difference. Only API version
+  would be bumped if the implementation changes.
+
+We should go with the Option C, i.e. use Terrajet in
+[`provider-aws`][provider-aws], provider-azure and provider-gcp as another way
+of generating CRDs and controllers so that they can reach the goal of coverage
+really fast. Then the community can keep investing in that single provider and
+make each CRD mature enough to depend on in production. In this way, we can
```

Comment:
Can `a consolidated way` be specified a little bit? I learned that the Alibaba Terraform provider team will focus on unit-test coverage. How will Terrajet increase the coverage?
Besides, the maturity might depend on some regression test in some way. How will all the e2e tests be composed for hundreds, even thousands of CRDs?

Reply:
The consolidation here refers to reducing number of providers for a cloud to 1 and note that the whole proposal is only about AWS, Azure and GCP. Alibaba provider maintainers or others can follow the track but they definitely don't have to.

> I learned that the Alibaba Terraform provider team will focus on unit-test coverage. How will Terrajet increase the coverage?

From that perspective, you can think of it like whatever investment you make on your Terraform provider will benefit you in Crossplane provider as well, including e2e and unit tests. For the providers mentioned here, the best we have today is to make sure examples work before we introduce the CRD and test against that if there is a change to that controller. I don't really expect much change in that regarding what tool is used to generate the CRD/controller.